tests/test.rs

```diff
@@ -151,6 +153,219 @@ async fn test_transaction() {
+            async move {
+                *transaction_attempted_count.lock().unwrap() += 1;
+
+                use user::dsl;
+                let _ = diesel::insert_into(dsl::user)
+                    .values((dsl::id.eq(0), dsl::name.eq("Jim")))
+                    .execute_async(&conn)
+                    .await?;
+                Ok(())
+            }
+        },
+        || async {
+            *should_retry_query_count.lock().unwrap() += 1;
+            true
+        },
+    )
+    .await
+    .expect("Transaction should have succeeded");
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+
+    // The transaction closure should have been attempted twice, but
+    // we should have only asked whether or not to retry once -- after
+    // the first failure, but not the second.
+    assert_eq!(
+        *transaction_attempted_count.lock().unwrap(),
+        EXPECTED_ERR_COUNT + 1
+    );
+    assert_eq!(
+        *should_retry_query_count.lock().unwrap(),
+        EXPECTED_ERR_COUNT
+    );
+
+    test_end(crdb).await;
+}
+
+#[tokio::test]
+async fn test_transaction_automatic_retry_does_not_retry_non_retryable_errors() {
+    let crdb = test_start().await;
+
+    let manager = async_bb8_diesel::ConnectionManager::<PgConnection>::new(&crdb.pg_config().url);
+    let pool = bb8::Pool::builder().build(manager).await.unwrap();
+    let conn = pool.get().await.unwrap();
+
+    // Test a transaction that:
+    //
+    // Fails with a non-retryable error. It should exit immediately.
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+    assert_eq!(
+        conn.transaction_async_with_retry(
+            |_| async { Err::<(), _>(diesel::result::Error::NotFound) },
+            || async { panic!("Should not attempt to retry this operation") }
+        )
+        .await
+        .expect_err("Transaction should have failed"),
+        diesel::result::Error::NotFound,
+    );
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+
+    test_end(crdb).await;
+}
+
+#[tokio::test]
+async fn test_transaction_automatic_retry_nested_transactions_fail() {
+    let crdb = test_start().await;
+
+    let manager = async_bb8_diesel::ConnectionManager::<PgConnection>::new(&crdb.pg_config().url);
+    let pool = bb8::Pool::builder().build(manager).await.unwrap();
+    let conn = pool.get().await.unwrap();
+
+    // This outer transaction should succeed immediately...
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+    conn.transaction_async_with_retry(
+        |conn| async move {
+            // ... but this inner transaction should fail! We do not support
+            // retryable nested transactions.
+            let err = conn
+                .transaction_async_with_retry(
+                    |_| async {
+                        panic!("Shouldn't run");
+                        #[allow(unreachable_code)]
+                        Ok(())
```

Comment:
Is this `Ok(())` needed for type inference inside this `async` block? A comment might help clarify.

Reply:
Exactly -- I'll add just that.